File: clients/python/games/go.py
from games.base import BaseGame
from framework.i18n import t
from typing import Optional, List


class Go(BaseGame):
    min_players = max_players = 2

    def __init__(self, size=9):
        self.size = size
        self.board = [[None] * size for _ in range(size)]
        self.players = []
        self._turn_idx = 0
        self._captures = {}
        self._prev_boards = set()
        self._passes = 0
        self._over = False
        self._winner = None

    def start(self, players):
        self.players = players
        self._captures = {p: 0 for p in players}

    def current_turn(self):
        return self.players[self._turn_idx] if self.players else None

    def _opponent(self, pid):
        return next(p for p in self.players if p != pid)

    def _neighbors(self, r, c):
        return [(r + dr, c + dc) for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]
                if 0 <= r + dr < self.size and 0 <= c + dc < self.size]

    def _group(self, r, c):
        color = self.board[r][c]
        if not color:
            return set(), set()
        visited, liberties = set(), set()
        stack = [(r, c)]
        while stack:
            cr, cc = stack.pop()
            if (cr, cc) in visited:
                continue
            visited.add((cr, cc))
            for nr, nc in self._neighbors(cr, cc):
                if self.board[nr][nc] == color:
                    stack.append((nr, nc))
                elif self.board[nr][nc] is None:
                    liberties.add((nr, nc))
        return visited, liberties

    def _remove_dead(self, color):
        removed = 0
        checked = set()
        for r in range(self.size):
            for c in range(self.size):
                if self.board[r][c] == color and (r, c) not in checked:
                    grp, libs = self._group(r, c)
                    checked |= grp
                    if not libs:
                        for gr, gc in grp:
                            self.board[gr][gc] = None
                        removed += len(grp)
        return removed

    def validate_move(self, player_id, move_data):
        if self._over or player_id != self.current_turn():
            return False
        if move_data.get('pass'):
            return True
        r, c = move_data.get('row'), move_data.get('col')
        if r is None or not (0 <= r < self.size and 0 <= c < self.size):
            return False
        if self.board[r][c] is not None:
            return False
        # Try move to check ko and suicide
        saved = [row[:] for row in self.board]
        self.board[r][c] = player_id
        opp = self._opponent(player_id)
        self._remove_dead(opp)
        _, libs = self._group(r, c)
        board_str = str(self.board)
        self.board = saved
        if board_str in self._prev_boards:
            return False
        if not libs:
            # Check if capture makes it valid
            self.board[r][c] = player_id
            self._remove_dead(opp)
            _, libs2 = self._group(r, c)
            self.board = saved
            if not libs2:
                return False
        return True

    def apply_move(self, player_id, move_data):
        if move_data.get('pass'):
            self._passes += 1
        else:
            self._passes = 0
            r, c = move_data['row'], move_data['col']
            self._prev_boards.add(str([row[:] for row in self.board]))
            self.board[r][c] = player_id
            opp = self._opponent(player_id)
            captured = self._remove_dead(opp)
            self._captures[player_id] = self._captures.get(player_id, 0) + captured
        if self._passes >= 2:
            self._over = True
            self._winner = self._score_winner()
        self._turn_idx = 1 - self._turn_idx

    def _score_winner(self):
        scores = {p: self._captures.get(p, 0) for p in self.players}
        for r in range(self.size):
            for c in range(self.size):
                if self.board[r][c]:
                    scores[self.board[r][c]] += 1
        scores[self.players[1]] += 6.5  # komi
        return max(scores, key=scores.get)

    def is_over(self):
        return self._over, self._winner

    def render(self, perspective=None):
        p0 = self.players[0] if self.players else 'black'
        p1 = self.players[1] if len(self.players) > 1 else 'white'
        syms = {None: '.', p0: 'B', p1: 'W'}
        lines = [t('go.title', size=self.size, captures=self._captures)]
        for row in self.board:
            lines.append(' '.join(syms.get(c, '.') for c in row))
        return '\n'.join(lines)

    def get_state(self, perspective=None) -> dict:
        return {
            'board': [row[:] for row in self.board],
            'turn': self.current_turn(),
            'players': self.players,
            'captures': dict(self._captures),
            'passes': self._passes,
        }

    def load_state(self, data: dict, perspective=None) -> None:
        if not data:
            return
        if 'players' in data:
            self.players = list(data['players'])
        if 'board' in data:
            self.board = [list(row) for row in data['board']]
        if 'turn' in data and data['turn'] in self.players:
            self._turn_idx = self.players.index(data['turn'])
        if 'captures' in data:
            self._captures = dict(data['captures'])
        if 'passes' in data:
            self._passes = data['passes']

    def parse_input(self, raw: str) -> Optional[dict]:
        import json as _json
        raw = raw.strip()
        if raw.lower() == 'pass':
            return {'pass': True}
        if raw.startswith('{'):
            try:
                obj = _json.loads(raw)
                if isinstance(obj, dict):
                    return obj
            except ValueError:
                pass
        parts = raw.split()
        if len(parts) == 2:
            try:
                return {'row': int(parts[0]), 'col': int(parts[1])}
            except ValueError:
                pass
        return None

    def get_help(self) -> List[str]:
        return [
            'Place stones to surround territory on a 9×9 board. Ko rule enforced.',
            'Higher score (territory + captures) wins.',
            'Move: <row> <col>   e.g. "3 4"   or   "pass"',
        ]

File: clients/node/src/games/go.js
'use strict';
const { BaseGame } = require('./base');
const { t } = require('../framework/i18n');

class Go extends BaseGame {
  constructor(size = 9) {
    super();
    this.size = size;
    this.board = Array.from({ length: size }, () => Array(size).fill(null));
    this.players = [];
    this._turnIdx = 0;
    this._captures = {};
    this._prevBoards = new Set();
    this._passes = 0;
    this._over = false;
    this._winner = null;
  }

  start(players) {
    this.players = players;
    this._captures = { [players[0]]: 0, [players[1]]: 0 };
  }

  currentTurn() {
    return this.players[this._turnIdx] || null;
  }

  _opponent(pid) {
    return this.players.find(p => p !== pid);
  }

  _neighbors(r, c) {
    return [[-1,0],[1,0],[0,-1],[0,1]]
      .map(([dr, dc]) => [r+dr, c+dc])
      .filter(([nr, nc]) => nr >= 0 && nr < this.size && nc >= 0 && nc < this.size);
  }

  _group(r, c) {
    const color = this.board[r][c];
    if (!color) return [new Set(), new Set()];
    const visited = new Set(), liberties = new Set();
    const stack = [[r, c]];
    while (stack.length) {
      const [cr, cc] = stack.pop();
      const key = `${cr},${cc}`;
      if (visited.has(key)) continue;
      visited.add(key);
      for (const [nr, nc] of this._neighbors(cr, cc)) {
        if (this.board[nr][nc] === color) stack.push([nr, nc]);
        else if (this.board[nr][nc] === null) liberties.add(`${nr},${nc}`);
      }
    }
    return [visited, liberties];
  }

  _groupCells(visited) {
    return [...visited].map(k => k.split(',').map(Number));
  }

  _removeDead(color) {
    let removed = 0;
    const checked = new Set();
    for (let r = 0; r < this.size; r++) {
      for (let c = 0; c < this.size; c++) {
        const key = `${r},${c}`;
        if (this.board[r][c] === color && !checked.has(key)) {
          const [grp, libs] = this._group(r, c);
          for (const k of grp) checked.add(k);
          if (!libs.size) {
            for (const [gr, gc] of this._groupCells(grp)) {
              this.board[gr][gc] = null;
              removed++;
            }
          }
        }
      }
    }
    return removed;
  }

  _boardStr() {
    return JSON.stringify(this.board);
  }

  _copyBoard() {
    return this.board.map(row => [...row]);
  }

  validateMove(playerId, moveData) {
    if (this._over || playerId !== this.currentTurn()) return false;
    if (moveData.pass) return true;
    const { row: r, col: c } = moveData;
    if (r == null || !(r >= 0 && r < this.size && c >= 0 && c < this.size)) return false;
    if (this.board[r][c] !== null) return false;
    const saved = this._copyBoard();
    this.board[r][c] = playerId;
    const opp = this._opponent(playerId);
    this._removeDead(opp);
    const [, libs] = this._group(r, c);
    const boardStr = this._boardStr();
    this.board = saved;
    if (this._prevBoards.has(boardStr)) return false;
    if (!libs.size) {
      this.board[r][c] = playerId;
      this._removeDead(opp);
      const [, libs2] = this._group(r, c);
      this.board = saved;
      if (!libs2.size) return false;
    }
    return true;
  }

  applyMove(playerId, moveData) {
    if (moveData.pass) {
      this._passes++;
    } else {
      this._passes = 0;
      const { row: r, col: c } = moveData;
      this._prevBoards.add(this._boardStr());
      this.board[r][c] = playerId;
      const opp = this._opponent(playerId);
      const captured = this._removeDead(opp);
      this._captures[playerId] = (this._captures[playerId] || 0) + captured;
    }
    if (this._passes >= 2) {
      this._over = true;
      this._winner = this._scoreWinner();
    }
    this._turnIdx = 1 - this._turnIdx;
  }

  _scoreWinner() {
    const scores = { ...this._captures };
    for (let r = 0; r < this.size; r++) {
      for (let c = 0; c < this.size; c++) {
        if (this.board[r][c]) scores[this.board[r][c]]++;
      }
    }
    scores[this.players[1]] = (scores[this.players[1]] || 0) + 6.5;
    return this.players.reduce((best, p) => (scores[p] > scores[best] ? p : best), this.players[0]);
  }

  isOver() {
    return [this._over, this._winner];
  }

  render(perspective) {
    const [p0, p1] = [this.players[0] || 'black', this.players[1] || 'white'];
    const syms = { [p0]: 'B', [p1]: 'W', null: '.' };
    const capStr = JSON.stringify(this._captures);
    const lines = [t('go.title', { size: this.size, captures: capStr })];
    for (const row of this.board) {
      lines.push(row.map(c => syms[c] || '.').join(' '));
    }
    return lines.join('\n');
  }

  getState(perspective) {
    return {
      board: this.board,
      turn: this.currentTurn(),
      captures: { ...this._captures },
      players: this.players,
    };
  }

  loadState(data) {
    if (!data) return;
    if (data.players) this.players = data.players;
    if (data.board) this.board = data.board.map(row => [...row]);
    if (data.captures) this._captures = { ...data.captures };
    if (data.turn != null) {
      const idx = this.players.indexOf(data.turn);
      this._turnIdx = idx >= 0 ? idx : 0;
    }
  }
}

module.exports = { Go };
